internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_events_test.go

```diff
@@ -58,137 +31,126 @@ func BenchmarkWatch(b *testing.B) {
+		},
+		{
+			name:     "onemessage",
+			cap:      100,
+			watchers: 10,
+			messages: 1,
+		},
+		{
+			name:     "manymessages_singlewatcher",
+			cap:      100,
+			watchers: 1,
+			messages: 50,
+		},
+		{
+			name:     "manymessages_manywatchers",
+			cap:      100,
+			watchers: 20,
+			messages: 50,
+		},
+		{
+			name:     "manymessages_overcap",
+			cap:      10,
+			watchers: 5,
+			messages: 200,
 		},
 	}
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			e := &Events{
-				subscribers: tt.fields.subscribers,
-			}
+			e := NewEvents(tt.cap)
 
 			var wg sync.WaitGroup
-			wg.Add(tt.count)
+			wg.Add(tt.watchers)
 
-			for i := 0; i < tt.count; i++ {
-				e.Watch(func(events <-chan machine.Event) { wg.Done() })
-			}
-
-			wg.Wait()
+			got := uint32(0)
 
-			// We need to lock here to prevent a race condition when checking the
-			// number of subscribers.
-			e.Lock()
-			defer e.Unlock()
-
-			// We can only check if the number of subscribers decreases because there
-			// is a race condition between the tear down of subscriber and the above
-			// lock. In other words, there is a chance that the number of subscribers
-			// is not zero.
-			if len(e.subscribers) > tt.count {
-				t.Errorf("Watch() = got %v subscribers, expected to be < %v", len(e.subscribers), tt.count)
-			}
-		})
-	}
-}
+			for i := 0; i < tt.watchers; i++ {
+				e.Watch(func(events <-chan machine.Event) {
+					defer wg.Done()
 
-func TestEvents_Publish(t *testing.T) {
-	type fields struct {
-		subscribers []chan machine.Event
-		Mutex       *sync.Mutex
-	}
+					for j := 0; j < tt.messages; j++ {
+						event := <-events
 
-	type args struct {
-		event proto.Message
-	}
+						var msg machine.SequenceEvent
 
-	tests := []struct {
-		name   string
-		count  int
-		fields fields
-		args   args
-	}{
-		{
-			name:  "success",
-			count: 10,
-			fields: fields{
-				subscribers: make([]chan machine.Event, 0, 100),
-				Mutex:       &sync.Mutex{},
-			},
-			args: args{
-				event: &machine.SequenceEvent{},
-			},
-		},
-	}
+						if err := proto.Unmarshal(event.GetData().GetValue(), &msg); err != nil {
+							t.Fatalf("failed to unmarshal message: %s", err)
+						}
 
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			e := &Events{
-				subscribers: tt.fields.subscribers,
-			}
-
-			var wg sync.WaitGroup
-			wg.Add(tt.count)
+						seq, err := strconv.Atoi(msg.Sequence)
+						if err != nil {
+							t.Fatalf("failed to convert sequence to number: %s", err)
+						}
 
-			mu := &sync.Mutex{}
+						if seq != j {
+							t.Fatalf("unexpected sequence: %d != %d", seq, j)
+						}
 
-			got := 0
+						atomic.AddUint32(&got, 1)
+					}
+				})
+			}
 
-			for i := 0; i < tt.count; i++ {
-				e.Watch(func(events <-chan machine.Event) {
-					<-events
+			l := rate.NewLimiter(1000, tt.cap/2)
 
-					mu.Lock()
-					got++
-					mu.Unlock()
+			for i := 0; i < tt.messages; i++ {
+				_ = l.Wait(context.Background())
```

Comment:
this is required to make publishing rate-limited, so that it doesn't overrun the buffer while watchers are still starting their goroutines